File: kglite/mcp_server/cypher_tools.py
"""YAML-declared `tools[].cypher` registration.

Mirrors `crates/kglite-mcp-server/src/cypher_tools.rs`. Each manifest
entry like

    tools:
      - name: session_detail
        description: Full source JSON for a session by id.
        cypher: |
          MATCH (s:Session {id: $session_id}) RETURN s
        parameters:
          type: object
          properties:
            session_id: { type: string }
          required: [session_id]

becomes an MCP tool that substitutes `$session_id` in the template
and dispatches to kglite's `graph.cypher(template, params=args)`. The
schema published with the tool is taken from `parameters:` verbatim
when present (must be a JSON-Schema object), otherwise a permissive
empty-object schema.
"""

from __future__ import annotations

from typing import Any

from mcp import types as mcp_types

from kglite.mcp_server.csv_http import CsvHttpConfig
from kglite.mcp_server.manifest import CypherTool
from kglite.mcp_server.tools import GraphState, run_cypher


def build_tool_attrs(tools: list[CypherTool]) -> list[mcp_types.Tool]:
    """Return mcp.Tool attrs for each manifest cypher tool."""
    attrs: list[mcp_types.Tool] = []
    for spec in tools:
        schema = spec.parameters if spec.parameters else {"type": "object", "properties": {}}
        attrs.append(
            mcp_types.Tool(
                name=spec.name,
                description=spec.description or f"Manifest-declared Cypher tool {spec.name!r}.",
                inputSchema=schema,
            )
        )
    return attrs


async def call_cypher_tool(
    spec: CypherTool,
    state: GraphState,
    args: dict[str, Any],
    csv_http: CsvHttpConfig | None,
) -> str:
    """Execute a manifest cypher template with arg substitution. Errors
    bubble up as text bodies (consistent with the Rust shim)."""
    return run_cypher(state, spec.cypher, params=args, csv_http=csv_http)

File: crates/kglite-mcp-server/src/cypher_tools.rs
//! YAML-declared `tools[].cypher` registration for kglite-mcp-server.
//!
//! mcp-methods 0.3.23 deliberately keeps the framework domain-agnostic
//! — it parses `ToolSpec::Cypher` entries from the manifest but doesn't
//! know how to run Cypher. We use the framework's now-public
//! `build_tool_attr` plus rmcp's `ToolRoute::new_dyn` directly to turn
//! each entry into a registered MCP tool whose handler dispatches
//! into the active graph's `cypher()` Python method.
//!
//! Per the 0.3.23 ack: "every domain-specific helper we add puts
//! pressure on the framework to know about query languages, runner
//! protocols, and graph-engine error shapes. None of that belongs
//! here." — so this module owns the boundary.

use std::pin::Pin;
use std::sync::Arc;

use anyhow::Result;
use mcp_methods::server::{build_tool_attr, Manifest, McpServer, ToolSpec};
use rmcp::handler::server::router::tool::ToolRoute;
use rmcp::handler::server::tool::ToolCallContext;
use rmcp::model::{CallToolResult, Content};
use rmcp::ErrorData as McpError;
use serde_json::{Map, Value};

use crate::tools::GraphState;

type DynFut<'a, T> = Pin<Box<dyn std::future::Future<Output = T> + Send + 'a>>;

/// Closure shape for executing a Cypher template with named arguments.
/// Receives the raw template string + the agent's argument map and
/// returns the rendered tool body (or an error string from the runner).
pub type CypherRunner =
    Arc<dyn Fn(&str, &Map<String, Value>) -> Result<String> + Send + Sync + 'static>;

/// Build a runner backed by the given `GraphState`. The runner forwards
/// to `graph.cypher(template, params=args)` on the active Python graph.
pub fn make_runner(state: GraphState) -> CypherRunner {
    Arc::new(move |template: &str, args: &Map<String, Value>| {
        Ok(state.run_cypher_template(template, args))
    })
}

/// Walk `manifest.tools` and register every `ToolSpec::Cypher` entry as
/// an MCP tool. Returns the number registered.
///
/// The schema published with each tool is taken from `parameters:` in
/// the YAML when present, falling back to an empty object schema. The
/// description comes from the entry's `description:` (otherwise empty).
pub fn register_cypher_tools(
    server: &mut McpServer,
    manifest: &Manifest,
    runner: CypherRunner,
) -> Result<usize> {
    let cypher_tools: Vec<_> = manifest
        .tools
        .iter()
        .filter_map(|t| match t {
            ToolSpec::Cypher(c) => Some(c),
            _ => None,
        })
        .collect();
    if cypher_tools.is_empty() {
        return Ok(0);
    }
    let count = cypher_tools.len();
    let router = server.tool_router_mut();
    for spec in cypher_tools {
        let schema = spec
            .parameters
            .as_ref()
            .and_then(|v| v.as_object().cloned())
            .unwrap_or_else(|| {
                let mut m = Map::new();
                m.insert("type".into(), Value::String("object".into()));
                m.insert("properties".into(), Value::Object(Map::new()));
                m
            });
        let attr = build_tool_attr(&spec.name, spec.description.as_deref(), schema);
        let template = spec.cypher.clone();
        let name = spec.name.clone();
        let runner = runner.clone();
        router.add_route(ToolRoute::new_dyn(
            attr,
            move |ctx: ToolCallContext<'_, McpServer>| -> DynFut<'_, Result<CallToolResult, McpError>> {
                let runner = runner.clone();
                let template = template.clone();
                let name = name.clone();
                let arguments = ctx.arguments.clone();
                Box::pin(async move {
                    let args: Map<String, Value> = arguments.unwrap_or_default();
                    let body = match runner(&template, &args) {
                        Ok(text) => text,
                        Err(e) => format!("cypher tool {name:?} error: {e}"),
                    };
                    Ok(CallToolResult::success(vec![Content::text(body)]))
                })
            },
        ));
    }
    Ok(count)
}
